lib/Sema/CSDiag.cpp

```diff
@@ -2315,19 +2315,6 @@ bool FailureDiagnosis::diagnoseContextualConversionError(
       exprType->isEqual(contextualType)) {
     return false;
   }
-  
-  // If we're trying to convert something of type "() -> T" to T, then we
-  // probably meant to call the value.
-  if (auto srcFT = exprType->getAs<AnyFunctionType>()) {
-    if (srcFT->getParams().empty() &&
-        !isUnresolvedOrTypeVarType(srcFT->getResult()) &&
-        CS.TC.isConvertibleTo(srcFT->getResult(), contextualType, CS.DC)) {
```

Comment:
Actually, it might be worth to temporarily replace `diagnose` here with `ContextualFailure::diagnoseMissingFunctionCall` but leave this check in CSDiag just in case.

Reply:
I cannot call it because it doesn't accept any arguments and I cannot create `ContextualFailure` here because the constructor is protected. I can change that however!

Reply:
I'll have to allow the function to accept arguments. Also, the constraint locator in some cases does not have any path elements, so if I create a fix and diagnose then it causes a crash. I guess I could just insert a ContextualType path element into the locator and then pass it?